code/modules/effects/map_effects/airlock_marker.dm

```diff
@@ -0,0 +1,135 @@
+	name = "map marker parent abstract object"
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "map_marker"
+
+/obj/effect/map_effect/marker_helper
+	name = "map marker helper parent abstract object"
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "map_marker"
+
+/// map of airlock marker `id_tag` to a list of map_effect/airlock_markers
+var/global/list/airlock_markers = list()
+
+/// Airlock marker that, when placed above airlock components (doors, pumps, sensors, etc),
+/// actually sets the airlock up to make it functional.
+/obj/effect/map_effect/marker/airlock
+	name = "airlock marker (inside the airlock)"
+	desc = "MASTER_TAG VAR MUST BE UNIQUE FOR THE AIRLOCK! Place this on top of airlock components (doors, pumps, sensors, etc)."
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "airlock_marker"
+	layer = LIGHTING_LAYER
+	/// Radio frequency of this airlock.
+	var/frequency = 2137
+	/// Unique tag for this airlock. Not visible in game and to the player. Do not leave this as null.
+	/// THIS MUST BE UNIQUE FOR THE AIRLOCK. Every marker in one airlock should have the same `master_tag`.
+	/// But different airlocks, even on different maps, cannot share the same `master_tag`.
+	var/master_tag = null
+	/// Doors/buttons/etc will be set to this access requirement. If null, they will not have any access requirements.
+	var/required_access = list(access_external_airlocks)
+
+/// Specialization helper for the airlock marker, to be put above "exterior" parts of the airlock,
+/// and on top of the actual airlock marker. By itself does nothing.
+/obj/effect/map_effect/marker_helper/airlock/exterior
+	name = "airlock marker (exterior/outside/vacuum)"
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "airlock_marker_exterior"
+	layer = LIGHTING_LAYER
+
+/// Specialization helper for the airlock marker, to be put above "interior" parts of the airlock,
+/// and on top of the actual airlock marker. By itself does nothing.
+/obj/effect/map_effect/marker_helper/airlock/interior
+	name = "airlock marker (interior/inside/pressurized)"
+	icon = 'icons/effects/map_effects.dmi'
+	icon_state = "airlock_marker_interior"
+	layer = LIGHTING_LAYER
+
+/// add the airlock market to `airlock_markers`
+/obj/effect/map_effect/marker/airlock/Initialize(mapload, ...)
+	..()
+	if(master_tag && frequency)
+		if(!airlock_markers[master_tag])
+			airlock_markers[master_tag] = list()
+		airlock_markers[master_tag] += src
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/effect/map_effect/marker/airlock/LateInitialize()
+	if(master_tag && frequency)
+		airlock_marker_init_airlock(master_tag)
+		airlock_markers[master_tag] = null // init only once
+
+#define MASTER_TAG		"[master_tag]_controller"
+#define AIRPUMP_TAG		"[master_tag]_pump"
+#define SENSOR_TAG		"[master_tag]_sensor"
+#define EXTERIOR_DOOR_TAG	"[master_tag]_outer"
+#define INTERIOR_DOOR_TAG	"[master_tag]_inner"
+
+/// actually set the airlock up
+/proc/airlock_marker_init_airlock(var/master_tag)
```

Comment:
Why is this a global proc, instead of a proc of `/obj/effect/map_effect/marker/airlock`?

Reply:
outdated/refactored